lnrpc/routerrpc/router_server.go

```diff
@@ -396,3 +407,81 @@ func marshallChannelUpdate(update *lnwire.ChannelUpdate) *ChannelUpdate {
 		FeeRate:         update.FeeRate,
 	}
 }
+func (s *Server) LookupPayment(request *LookupPaymentRequest,
+	stream Router_LookupPaymentServer) error {
+
+	paymentHash, err := lntypes.MakeHash(request.PaymentHash)
+	if err != nil {
+		return err
+	}
+
+	log.Debugf("LookupPayment called for payment %v", paymentHash)
+
+	// Subscribe to the outcome of this payment.
+	inFlight, events, err := s.cfg.RouterBackend.Tower.SubscribePayment(
+		paymentHash,
+	)
+	switch err {
+	case nil:
+	case channeldb.ErrPaymentNotInitiated:
+		return status.Error(codes.NotFound, err.Error())
+	default:
+		return err
+	}
+
+	// If it is in flight, send a state update to the client. LookupPayment
+	// is expected to always send the current payment state immediately.
+	if inFlight {
+		err = stream.Send(&PaymentStatus{
+			State: PaymentState_IN_FLIGHT,
+		})
+		if err != nil {
+			return err
+		}
+	}
+
+	// Wait for the outcome of the payment. For payments that have already
+	// completed, the event should already be waiting on the channel.
+	select {
+	case event := <-events:
+		// Marshall event to rpc type.
+		var status PaymentStatus
+
+		switch event.Status {
+
+		case channeldb.StatusFailed:
+			switch event.FailureReason {
+
+			case channeldb.FailureReasonTimeout:
+				status.State = PaymentState_FAILED_TIMEOUT
+
+			case channeldb.FailureReasonNoRoute:
+				status.State = PaymentState_FAILED_NO_ROUTE
+
+			default:
+				return errors.New("unknown failure reason")
```

Comment:
I feel like we're losing _too much_ context here. For example, without unknown payment hash, this this API can't really be used for probing. Alternatively, without the final expiry too soon error, then it makes it harder to debug issues like the sending node being behind the main chain.

Reply:
Yes, that information is definitely useful. Terminal errors need to be exposed. See comment above about follow up pr.